File: app/src/main/java/com/ocram/qichwadic/data/datasource/DictionaryCloudDataStore.java
package com.ocram.qichwadic.data.datasource;

import com.ocram.qichwadic.data.repository.DictionaryRepositoryImpl;
import com.ocram.qichwadic.domain.model.Definition;
import com.ocram.qichwadic.domain.model.Dictionary;
import com.ocram.qichwadic.framework.net.client.RetrofitClient;

import java.util.List;

import javax.inject.Inject;

import io.reactivex.Flowable;

public class DictionaryCloudDataStore implements DictionaryRepositoryImpl.CloudDataStore {

    private final RetrofitClient retrofitClient;

    @Inject
    public DictionaryCloudDataStore(RetrofitClient retrofitClient) {
        this.retrofitClient = retrofitClient;
    }

    @Override
    public Flowable<List<Dictionary>> getDictionaries() {
        return retrofitClient.getDictionaries();
    }

    @Override
    public Flowable<List<Definition>> getDefinitionsByDictionary(String entriesUrl) {
        return retrofitClient.getAllDefinitionsByDictionary(entriesUrl);
    }
}

File: app/src/main/java/com/ocram/qichwadic/features/dictionaries/data/datastore/DictionaryCloudDataStore.kt
package com.ocram.qichwadic.features.dictionaries.data.datastore

import com.ocram.qichwadic.core.data.remote.ApiResponse
import com.ocram.qichwadic.core.data.remote.CloudDataStore
import com.ocram.qichwadic.core.data.model.DefinitionEntity
import com.ocram.qichwadic.core.data.model.DictionaryEntity
import com.ocram.qichwadic.core.data.remote.RetrofitClient

interface DictionaryCloudDataStore {

    suspend fun getDictionaries(): ApiResponse<out List<DictionaryEntity>>

    suspend fun getDefinitionsByDictionary(dictionaryId: Int): ApiResponse<out List<DefinitionEntity>>
}

class DictionaryCloudDataStoreImpl
constructor(private val retrofitClient: RetrofitClient): CloudDataStore(), DictionaryCloudDataStore {

    override suspend fun getDictionaries(): ApiResponse<out List<DictionaryEntity>> {
        return processResponse(retrofitClient.getDictionaries())
    }

    override suspend fun getDefinitionsByDictionary(dictionaryId: Int): ApiResponse<out List<DefinitionEntity>> {
        return processResponse(retrofitClient.getAllDefinitionsByDictionary(dictionaryId))
    }
}
